esp/esp/program/modules/handlers/programprintables.py

```diff
@@ -888,23 +887,46 @@ def studentschedules(self, request, tl, one, two, module, extra, prog, onsite=Fa
     def get_student_schedules(request, students, prog, extra='', onsite=False):
         """ generate student schedules """
         context = {}
+
+        # to avoid a query per student, get all the classes and SRs upfront
+        all_classes = ClassSection.objects.filter(
+            nest_Q(StudentRegistration.is_valid_qobject(),
+                   'studentregistration'),
+            studentregistration__user__in=students,
+            studentregistration__relationship__name='Enrolled',
+            parent_class__parent_program=prog,
+            status=ACCEPTED,
+            meeting_times__isnull=False).distinct()
+        all_classes = all_classes.select_related('parent_class')
+        all_classes = all_classes.prefetch_related('meeting_times')
+        classes_by_id = {cls.id: cls for cls in all_classes}
+
+        sr_pairs = all_classes.values_list('id', 'studentregistration__user')
+        classes_by_student = collections.defaultdict(list)
+        for cls_id, user_id in sr_pairs:
+            classes_by_student[user_id].append(classes_by_id[cls_id])
+
+        for user_id in classes_by_student:
+            # Sort the classes.  We don't want to use 
+            classes_by_student[user_id].sort(
+                key=lambda cls: (cls.start_time_prefetchable(), cls.title()))
+
+        times_compulsory = Event.objects.filter(program=prog, event_type__description='Compulsory').order_by('start')
+        for t in times_compulsory:
```

Comment:
(We talked in person, but FTSOTR, I think it's roughly trying to fake a class that the template will accept for a compulsory timeblock, such as a required lunch.)